neqo-transport/src/send_stream.rs

```diff
@@ -602,20 +571,43 @@ impl SendStream {
     }
 
     pub fn send(&mut self, buf: &[u8]) -> Res<usize> {
-        let sent = match self.state {
-            SendStreamState::Ready => {
-                let mut send_buf = TxBuffer::new();
-                let sent = send_buf.send(buf);
-                self.state.transition(SendStreamState::Send { send_buf });
-                sent
+        if let SendStreamState::Ready = self.state {
+            self.state.transition(SendStreamState::Send {
+                send_buf: TxBuffer::new(),
+            });
+        }
+
+        let stream_credit_avail = self.credit_avail();
+        let conn_credit_avail = self.flow_mgr.borrow().conn_credit_avail();
+        let credit_avail = min(stream_credit_avail, conn_credit_avail);
+        let can_send_bytes = min(credit_avail, buf.len() as u64);
+
+        if !buf.is_empty() && can_send_bytes == 0 {
```

Comment:
We should assert that buf is non-empty at the start of all of this.  There is no reason to pass `&[]` into this function.

Reply:
Concur, but since this would be due to an error in the caller rather than in this code, I think returning an error and emitting an error-level log message is appropriate.